plugins/uefi-dbx/fu-uefi-dbx-device.c

```diff
@@ -0,0 +1,153 @@
+
+	/* get default image */
+	fw = fu_firmware_get_image_default_bytes (firmware, error);
+	if (fw == NULL)
+		return FALSE;
+
+	/* write entire chunk to efivarfs */
+	fu_device_set_status (device, FWUPD_STATUS_DEVICE_WRITE);
+	buf = g_bytes_get_data (fw, &bufsz);
+	if (!fu_efivar_set_data (FU_EFIVAR_GUID_SECURITY_DATABASE,
+				 "dbx", buf, bufsz,
+				 FU_EFIVAR_ATTR_APPEND_WRITE |
+				 FU_EFIVAR_ATTR_TIME_BASED_AUTHENTICATED_WRITE_ACCESS |
+				 FU_EFIVAR_ATTR_RUNTIME_ACCESS |
+				 FU_EFIVAR_ATTR_BOOTSERVICE_ACCESS |
+				 FU_EFIVAR_ATTR_NON_VOLATILE,
+				 error)) {
+		return FALSE;
+	}
+
+	/* success! */
+	return TRUE;
+}
+
+static gboolean
+fu_uefi_dbx_device_set_version_number (FuDevice *device, GError **error)
+{
+	gsize bufsz = 0;
+	guint csum_cnt = 0;
+	g_autofree gchar *version = NULL;
+	g_autofree guint8 *buf = NULL;
+	g_autoptr(GPtrArray) dbx = NULL;
+	const gchar *ignored_guids[] = {
+		FU_EFI_SIGNATURE_GUID_OVMF,
+		NULL };
+
+	/* use the number of checksums in the dbx as a version number, ignoring
+	 * some owners that do not make sense */
+	if (!fu_efivar_get_data (FU_EFIVAR_GUID_SECURITY_DATABASE, "dbx",
+				 &buf, &bufsz, NULL, error))
+		return FALSE;
+	dbx = fu_efi_signature_parser_new (buf, bufsz,
+					   FU_EFI_SIGNATURE_PARSER_FLAGS_NONE,
+					   error);
+	if (dbx == NULL)
+		return FALSE;
+	for (guint j = 0; j < dbx->len; j++) {
+		FuEfiSignatureList *siglist = g_ptr_array_index (dbx, j);
+		GPtrArray *items = fu_efi_signature_list_get_all (siglist);
+		for (guint i = 0; i < items->len; i++) {
+			FuEfiSignature *sig = g_ptr_array_index (items, i);
+			if (fu_efi_signature_get_kind (sig) != FU_EFI_SIGNATURE_KIND_SHA256)
+				continue;
+			if (g_strv_contains (ignored_guids, fu_efi_signature_get_owner (sig)))
+				continue;
+			csum_cnt++;
+		}
+	}
+	version = g_strdup_printf ("%u", csum_cnt);
+	fu_device_set_version (device, version);
+	fu_device_set_version_lowest (device, version);
+	return TRUE;
+}
+
+static gboolean
+fu_uefi_dbx_device_probe (FuDevice *device, GError **error)
+{
+	gsize bufsz = 0;
+	g_autofree gchar *devid = NULL;
+	g_autofree guint8 *buf = NULL;
+	g_autoptr(GPtrArray) kek = NULL;
+
+	/* use each of the certificates in the KEK to generate the GUIDs */
+	if (!fu_efivar_get_data (FU_EFIVAR_GUID_EFI_GLOBAL, "KEK",
+				 &buf, &bufsz, NULL, error))
+		return FALSE;
+	kek = fu_efi_signature_parser_new (buf, bufsz,
+					   FU_EFI_SIGNATURE_PARSER_FLAGS_NONE,
+					   error);
+	if (kek == NULL)
+		return FALSE;
+	for (guint i = 0; i < kek->len; i++) {
+		FuEfiSignatureList *siglist = g_ptr_array_index (kek, i);
+		GPtrArray *sigs = fu_efi_signature_list_get_all (siglist);
+		for (guint j = 0; j < sigs->len; j++) {
+			FuEfiSignature *sig = g_ptr_array_index (sigs, j);
+			fu_device_add_instance_id (device, fu_efi_signature_get_checksum (sig));
+		}
+	}
+	return fu_uefi_dbx_device_set_version_number (device, error);
+}
+
+static void
+fu_uefi_dbx_device_init (FuUefiDbxDevice *self)
```

Comment:
I think it makes sense to set the parent of this device as system firmware

Reply:
Good idea, fixed.